Experimental async support (#629)
neo4j/neo4j-python-driver: docs/source/async_api.rst

.. _async-api-documentation:

#######################
Async API Documentation
#######################

.. warning::
    The whole async API is currently in experimental phase.

>     This means everything documented on this page might be removed or change

Reviewer: ```suggestion
    This means everything documented on this page might be removed or changed
```

>     its API at any time (including in patch releases).

Reviewer: ```suggestion
    at any time (including in patch releases).
```

******************
AsyncGraphDatabase
******************

Async Driver Construction
=========================

The :class:`neo4j.AsyncDriver` construction is done via a `classmethod` on the :class:`neo4j.AsyncGraphDatabase` class.

.. autoclass:: neo4j.AsyncGraphDatabase
   :members: driver

> Example, driver creation:

Reviewer: ```suggestion
Driver creation example:
```

.. code-block:: python

    import asyncio

    from neo4j import AsyncGraphDatabase

    async def main():
        uri = "neo4j://example.com:7687"
        driver = AsyncGraphDatabase.driver(uri, auth=("neo4j", "password"))

        await driver.close()  # close the driver object

     asyncio.run(main())

> For basic authentication, ``auth`` can be a simple tuple, for example:

Reviewer: ```suggestion
For basic authentication, `auth` can be a simple tuple, for example:
```

.. code-block:: python

   auth = ("neo4j", "password")

This will implicitly create a :class:`neo4j.Auth` with a ``scheme="basic"``.
Other authentication methods are described under :ref:`auth-ref`.

> Example, with block context:

Reviewer: ```suggestion
`with` block context example:
```

            yield "::1", 7687
            yield "127.0.0.1", 7687
        else:
            from socket import gaierror
            raise gaierror("Unexpected socket address %r" % socket_address)

    # alternatively
    def custom_resolver(socket_address):
        ...

    driver = AsyncGraphDatabase.driver("neo4j://example.com:9999",
                                       auth=("neo4j", "password"),
                                       resolver=custom_resolver)

:Default: ``None``

Driver Object Lifetime
======================

For general applications, it is recommended to create one top-level :class:`neo4j.AsyncDriver` object that lives for the lifetime of the application.

For example:

.. code-block:: python

    from neo4j import AsyncGraphDatabase

    class Application:

        def __init__(self, uri, user, password)
            self.driver = AsyncGraphDatabase.driver(uri, auth=(user, password))

        async def close(self):
            await self.driver.close()

Connection details held by the :class:`neo4j.AsyncDriver` are immutable.
Therefore if, for example, a password is changed, a replacement :class:`neo4j.AsyncDriver` object must be created.

> More than one :class:`.AsyncDriver` may be required if connections to multiple databases, or connections as multiple users, are required.

Reviewer: maybe mention impersonation here, since multiple users can now be configured at the session level

:class:`neo4j.AsyncDriver` objects are safe to be used in concurrent coroutines.
They are not thread-safe.

.. _async-bolt-driver-ref:

AsyncBoltDriver
===============

URI schemes:
    ``bolt``, ``bolt+ssc``, ``bolt+s``

Will result in:

.. autoclass:: neo4j.AsyncBoltDriver

.. _async-neo4j-driver-ref:

AsyncNeo4jDriver
================

URI schemes:
    ``neo4j``, ``neo4j+ssc``, ``neo4j+s``

Will result in:

.. autoclass:: neo4j.AsyncNeo4jDriver

*********************************
AsyncSessions & AsyncTransactions
*********************************

> All database activity is co-ordinated through two mechanisms: the :class:`neo4j.AsyncSession` and the :class:`neo4j.AsyncTransaction`.

Reviewer: ```suggestion
All database activity is coordinated through two mechanisms: the :class:`neo4j.AsyncSession` and the :class:`neo4j.AsyncTransaction`.
```

A :class:`neo4j.AsyncSession` is a logical container for any number of causally-related transactional units of work.
Sessions automatically provide guarantees of causal consistency within a clustered environment but multiple sessions can also be causally chained if required.

> Sessions provide the top-level of containment for database activity.

Reviewer: missing word after "top-level" or extra hyphen?
